File: backend/src/routes/analysis.js
// @ts-check

import { Router } from 'express';
import { analyzeAlerts, chat } from '../services/llm.js';
import { asyncHandler } from '../middleware/asyncHandler.js';
import { getConfig, getConfigInt } from '../services/config.js';

const ENV_AGENT_SERVICE_URL = process.env.AGENT_SERVICE_URL || 'http://localhost:8002';

/** 动态获取 Agent 服务地址（优先数据库，回退环境变量） */
function getAgentUrl() {
  return getConfig('agent.service_url', ENV_AGENT_SERVICE_URL);
}

/**
 * @typedef {{ trace_id: string, status?: string, message?: string }} AgentServiceResponse
 */

/**
 * @param {unknown} error
 * @returns {string}
 */
function getErrorMessage(error) {
  return error instanceof Error ? error.message : String(error);
}

/**
 * @param {number[]} requestedIds
 * @param {{ id: number }[]} alerts
 * @returns {number[]}
 */
function findMissingAlertIds(requestedIds, alerts) {
  const foundIds = new Set(alerts.map((alert) => Number(alert.id)));
  return requestedIds.filter((id) => !foundIds.has(id));
}

/**
 * 调用 Python Agent Service 进行分析
 * @param {number[]} alertIds
 * @param {{ fetchFn?: typeof fetch, agentServiceUrl?: string }} [options]
 * @returns {Promise<AgentServiceResponse>}
 */
async function callAgentService(alertIds, { fetchFn = fetch, agentServiceUrl = getAgentUrl() } = {}) {
  const resp = await fetchFn(`${agentServiceUrl}/analyze`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ alert_ids: alertIds }),
  });
  if (!resp.ok) {
    throw new Error(`Agent Service returned ${resp.status}`);
  }
  return /** @type {Promise<AgentServiceResponse>} */ (resp.json());
}

/**
 * 获取 Python Agent Service 状态
 * @param {{ fetchFn?: typeof fetch, agentServiceUrl?: string }} [options]
 * @returns {Promise<any>}
 */
async function getAgentServiceStatus({ fetchFn = fetch, agentServiceUrl = getAgentUrl() } = {}) {
  const resp = await fetchFn(`${agentServiceUrl}/status`, {
    method: 'GET',
    headers: { Accept: 'application/json' },
  });
  if (!resp.ok) {
    throw new Error(`Agent Service returned ${resp.status}`);
  }
  return resp.json();
}

/**
 * POST /api/analysis/alerts
 * AI 分析告警：委托 Python Agent Service 进行分析
 *
 * Body: { alert_ids: [1, 2, 3] }
 */
export function createAnalysisRouter({
  fetchFn = fetch,
  chatFn = chat,
  agentServiceUrl = getAgentUrl(),
} = {}) {
  const router = Router();

  router.get('/agent/status', asyncHandler(async (_req, res) => {
    try {
      const status = await getAgentServiceStatus({ fetchFn, agentServiceUrl });
      return res.json(status);
    } catch (agentErr) {
      console.error('Agent status unavailable:', getErrorMessage(agentErr));
      return res.status(503).json({ error: 'Agent Service unavailable' });
    }
  }));

  router.get('/mcp/servers', asyncHandler(async (_req, res) => {
    try {
      const resp = await fetchFn(`${agentServiceUrl}/mcp/servers`, {
        method: 'GET',
        headers: { Accept: 'application/json' },
      });
      if (!resp.ok) {
        throw new Error(`Agent Service returned ${resp.status}`);
      }
      const data = await resp.json();
      return res.json(data);
    } catch (err) {
      console.error('MCP servers info unavailable:', getErrorMessage(err));
      return res.status(503).json({ error: 'Agent Service unavailable' });
    }
  }));

  router.post('/alerts', asyncHandler(async (/** @type {any} */ req, /** @type {any} */ res) => {
    const { alert_ids } = req.body;
    if (!Array.isArray(alert_ids) || alert_ids.length === 0) {
      return res.status(400).json({ error: 'alert_ids[] required' });
    }

    if (!alert_ids.every((id) => Number.isInteger(id) && id > 0)) {
      return res.status(400).json({ error: 'alert_ids must contain positive integers' });
    }

    const db = req.db;
    const placeholders = alert_ids.map(() => '?').join(',');
    const alerts = db.prepare(`SELECT * FROM alerts WHERE id IN (${placeholders})`).all(...alert_ids);

    if (alerts.length === 0) {
      return res.status(404).json({ error: 'No alerts found' });
    }

    const missingIds = findMissingAlertIds(alert_ids, alerts);
    if (missingIds.length > 0) {
      return res.status(404).json({ error: 'Some alerts were not found', missing_alert_ids: missingIds });
    }

    try {
      const agentResult = await callAgentService(alert_ids, { fetchFn, agentServiceUrl });
      const updateStatus = db.prepare('UPDATE alerts SET status = ? WHERE id = ?');
      for (const id of alert_ids) {
        updateStatus.run('analyzing', id);
      }

      return res.json({
        trace_id: agentResult.trace_id,
        status: 'analyzing',
        message: 'Analysis delegated to Agent Service',
      });
    } catch (agentErr) {
      console.error('Agent Service unavailable:', getErrorMessage(agentErr));
      return res.status(503).json({ error: 'Agent Service unavailable' });
    }
  }));

/**
 * 从数据库获取当前系统上下文，注入给 LLM
 * @param {any} db
 * @returns {string}
 */
  function buildSystemContext(db) {
    try {
      const parts = [];

      // 告警统计
      const stats = db.prepare(`
        SELECT
          COUNT(*) as total,
          SUM(CASE WHEN status = 'open' THEN 1 ELSE 0 END) as open_count,
          SUM(CASE WHEN status = 'analyzing' THEN 1 ELSE 0 END) as analyzing_count,
          SUM(CASE WHEN severity = 'critical' THEN 1 ELSE 0 END) as critical_count,
          SUM(CASE WHEN severity = 'high' THEN 1 ELSE 0 END) as high_count,
          SUM(CASE WHEN severity = 'medium' THEN 1 ELSE 0 END) as medium_count,
          SUM(CASE WHEN severity = 'low' THEN 1 ELSE 0 END) as low_count
        FROM alerts
      `).get();
      if (stats && stats.total > 0) {
        parts.push(`告警概览: 共${stats.total}条告警, 待处理${stats.open_count}条, 分析中${stats.analyzing_count}条`);
        parts.push(`严重度分布: 严重${stats.critical_count} / 高${stats.high_count} / 中${stats.medium_count} / 低${stats.low_count}`);
      } else {
        parts.push('告警概览: 当前系统无告警记录');
      }

      // 最近 5 条告警
      const recentAlerts = db.prepare(`
        SELECT id, title, severity, source, src_ip, dst_ip, status, created_at
        FROM alerts ORDER BY created_at DESC LIMIT 5
      `).all();
      if (recentAlerts.length > 0) {
        parts.push('\n最近告警:');
        for (const a of recentAlerts) {
          parts.push(`  - [${a.severity}] ${a.title} (来源:${a.source}, ${a.src_ip||'?'}→${a.dst_ip||'?'}, 状态:${a.status})`);
        }
      }

      // 攻击链
      const chains = db.prepare(`
        SELECT id, name, stage, confidence FROM attack_chains ORDER BY created_at DESC LIMIT 3
      `).all();
      if (chains.length > 0) {
        parts.push('\n最近攻击链:');
        for (const c of chains) {
          parts.push(`  - ${c.name} (阶段:${c.stage}, 置信度:${c.confidence})`);
        }
      }

      // 资产
      const assetCount = db.prepare('SELECT COUNT(*) as cnt FROM assets').get();
      if (assetCount && assetCount.cnt > 0) {
        parts.push(`\n资产: 共${assetCount.cnt}台受监控资产`);
      }

      return parts.join('\n');
    } catch (err) {
      console.error('Failed to build system context:', err);
      return '';
    }
  }

/**
 * POST /api/analysis/chat
 * AI 对话：指挥面板中的自由问答，自动注入系统上下文
 *
 * Body: { messages: [{ role: 'user', content: '...' }] }
 */
  router.post('/chat', asyncHandler(async (/** @type {any} */ req, /** @type {any} */ res) => {
    const { messages } = req.body;
    if (!Array.isArray(messages)) {
      return res.status(400).json({ error: 'messages[] required' });
    }
    const maxMessages = getConfigInt('chat.max_messages', 50);
    if (messages.length > maxMessages) {
      return res.status(400).json({ error: `messages[] exceeds max length (${maxMessages})` });
    }
    try {
      const systemContext = buildSystemContext(req.db);
      const reply = await chatFn(messages, systemContext);
      res.json({ reply });
    } catch (err) {
      console.error('Chat failed:', getErrorMessage(err));
      res.status(500).json({ error: 'Chat service temporarily unavailable' });
    }
  }));

/**
 * GET /api/analysis/chains
 * 查询攻击链列表
 */
  router.get('/chains', (/** @type {any} */ req, /** @type {any} */ res) => {
    const db = req.db;
    const rawChains = db.prepare(`
      SELECT c.*
      FROM attack_chains c
      ORDER BY c.created_at DESC
      LIMIT 50
    `).all();

    const getAlerts = db.prepare(`SELECT id, title, severity, src_ip FROM alerts WHERE id IN (SELECT value FROM json_each(?))`);
    const getDecisions = db.prepare(`SELECT * FROM decisions WHERE attack_chain_id = ? ORDER BY created_at DESC, id DESC`);

    const chains = rawChains.map((/** @type {any} */ c) => {
      let alertIds = [];
      try { alertIds = JSON.parse(c.alert_ids || '[]'); } catch { /* ignore */ }
      const alerts = alertIds.length > 0 ? getAlerts.all(c.alert_ids) : [];

      const decisions = getDecisions.all(c.id);

      const alert_count = alertIds.length;
      const pendingDecisions = decisions.filter((/** @type {any} */ d) => d.status === 'pending');
      const status = decisions.length === 0 ? 'new' : pendingDecisions.length > 0 ? 'pending' : 'resolved';

      const primaryDecision = decisions[0] || null;
      const risk_level = primaryDecision?.risk_level || c.risk_level || 'medium';
      const recommendation = primaryDecision?.recommendation || null;
      const action_type = primaryDecision?.action_type || null;
      const decision_status = primaryDecision?.status || null;

      return {
        ...c,
        alerts,
        decisions: decisions.map((decision) => ({
          ...decision,
          action: decision.recommendation,
        })),
        alert_count,
        status,
        risk_level,
        recommendation,
        action_type,
        decision_status,
      };
    });

    res.json({ chains });
  });

/**
 * PATCH /api/analysis/decisions/:id
 * 人工反馈：接受或拒绝处置建议
 * Uses transaction for atomicity.
 */
  router.patch('/decisions/:id', (/** @type {any} */ req, /** @type {any} */ res) => {
    const { status } = req.body;
    if (!['accepted', 'rejected'].includes(status)) {
      return res.status(400).json({ error: 'status must be accepted or rejected' });
    }
    const db = req.db;

    // Atomic update: only succeeds if decision is still pending
    const result = db.prepare(
      'UPDATE decisions SET status = ? WHERE id = ? AND status = ?'
    ).run(status, req.params.id, 'pending');

    if (result.changes === 0) {
      const exists = db.prepare('SELECT id, status FROM decisions WHERE id = ?').get(req.params.id);
      if (!exists) {
        return res.status(404).json({ error: 'Decision not found' });
      }
      return res.status(400).json({ error: `Decision already processed (status: ${exists.status})` });
    }

    // Update related alert statuses in a transaction
    const updateRelatedAlerts = db.transaction(() => {
      const decision = db.prepare('SELECT attack_chain_id FROM decisions WHERE id = ?').get(req.params.id);
      if (decision) {
        const chain = db.prepare('SELECT alert_ids FROM attack_chains WHERE id = ?').get(decision.attack_chain_id);
        if (chain && chain.alert_ids) {
          try {
            const alertIds = JSON.parse(chain.alert_ids);
            const updateAlert = db.prepare('UPDATE alerts SET status = ? WHERE id = ?');
            const newAlertStatus = status === 'accepted' ? 'resolved' : 'analyzed';
            for (const alertId of alertIds) {
              updateAlert.run(newAlertStatus, alertId);
            }
          } catch { /* ignore malformed alert_ids */ }
        }
      }

      // Write audit log
      db.prepare(
        `INSERT INTO audit_logs (trace_id, alert_id, event_type, data) VALUES (?, ?, ?, ?)`
      ).run(
        `decision-${req.params.id}`,
        '',
        'decision_feedback',
        JSON.stringify({ decision_id: req.params.id, status })
      );
    });

    try {
      updateRelatedAlerts();
    } catch (err) {
      console.error('Failed to update related alerts:', err);
      // Decision itself was updated, but related alert updates failed
    }

    res.json({ updated: true });
  });

  return router;
}

export default createAnalysisRouter();

File: agent/routers/analysis.py
"""
Analysis router — agent status, alert analysis, chat, attack chains, decisions.

Migrated from Express backend/src/routes/analysis.js
Existing /analyze and /chat endpoints from service.py are preserved in service.py;
this router adds the Express-facing /api/analysis/* endpoints.
"""

from __future__ import annotations

import json
import logging
import os
from typing import Any, Optional

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field

from agent.routers._db import get_db

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/analysis", tags=["analysis"])

AGENT_SERVICE_URL = os.environ.get("AGENT_SERVICE_URL", "http://localhost:8002")


# ---------------------------------------------------------------------------
# Config helpers
# ---------------------------------------------------------------------------

def _get_config(conn, key: str, default: str = "") -> str:
    row = conn.execute("SELECT value FROM system_config WHERE key = ?", (key,)).fetchone()
    return row["value"] if row else default


def _get_config_int(conn, key: str, default: int = 0) -> int:
    raw = _get_config(conn, key)
    if not raw:
        return default
    try:
        return int(raw)
    except (ValueError, TypeError):
        return default


# ---------------------------------------------------------------------------
# Pydantic models
# ---------------------------------------------------------------------------

class AnalysisAlertRequest(BaseModel):
    alert_ids: list[int] = Field(..., min_length=1)


class ChatMessageItem(BaseModel):
    role: str
    content: str


class AnalysisChatRequest(BaseModel):
    messages: list[ChatMessageItem]


class DecisionFeedbackRequest(BaseModel):
    status: str = Field(..., pattern=r"^(accepted|rejected)$")


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------

@router.get("/agent/status")
async def agent_status(request: Request):
    """Proxy agent status — returns status from the local FastAPI service."""
    # Since we ARE the agent service now, just call the status logic directly
    from agent.service import _mcp_client, _running_tasks

    mcp_connected = _mcp_client is not None
    mcp_servers = _mcp_client.get_connected_servers() if _mcp_client else []
    active_tasks = sum(1 for t in _running_tasks.values() if t["status"] in ("started", "running"))

    return {
        "status": "ok",
        "mcp_connected": mcp_connected,
        "mcp_servers": mcp_servers,
        "running_tasks": active_tasks,
    }


@router.get("/mcp/servers")
async def mcp_servers_proxy():
    """Proxy MCP servers info — returns MCP server data directly."""
    from agent.service import mcp_servers as _mcp_servers_handler
    return await _mcp_servers_handler()


@router.post("/alerts")
async def analyze_alerts(body: AnalysisAlertRequest):
    """Trigger alert analysis via the agent service."""
    alert_ids = body.alert_ids

    if not all(isinstance(aid, int) and aid > 0 for aid in alert_ids):
        raise HTTPException(status_code=400, detail="alert_ids must contain positive integers")

    conn = get_db()
    try:
        placeholders = ",".join("?" for _ in alert_ids)
        alerts = conn.execute(f"SELECT * FROM alerts WHERE id IN ({placeholders})", alert_ids).fetchall()

        if not alerts:
            raise HTTPException(status_code=404, detail="No alerts found")

        found_ids = {a["id"] for a in alerts}
        missing = [aid for aid in alert_ids if aid not in found_ids]
        if missing:
            raise HTTPException(status_code=404, detail=f"Some alerts were not found: missing_alert_ids={missing}")

        # Call the analyze endpoint directly (we are the agent service)
        from agent.service import AnalyzeRequest, analyze as _analyze
        result = await _analyze(AnalyzeRequest(alert_ids=alert_ids))

        # Update alert statuses to "analyzing"
        for aid in alert_ids:
            conn.execute("UPDATE alerts SET status = ? WHERE id = ?", ("analyzing", aid))
        conn.commit()

        return {
            "trace_id": result.trace_id,
            "status": "analyzing",
            "message": "Analysis delegated to Agent Service",
        }
    except HTTPException:
        raise
    except Exception as exc:
        logger.exception("Agent Service error")
        raise HTTPException(status_code=503, detail="Agent Service unavailable") from exc


@router.post("/chat")
async def analysis_chat(body: AnalysisChatRequest):
    """AI chat with system context injection."""
    conn = get_db()
    try:
        max_messages = _get_config_int(conn, "chat.max_messages", 50)
        if len(body.messages) > max_messages:
            raise HTTPException(status_code=400, detail=f"messages[] exceeds max length ({max_messages})")

        system_context = _build_system_context(conn)

        # Use the LLM directly (ported from Express chat logic)
        from openai import AsyncOpenAI
        from agent.agent import DEFAULT_MODEL, DEFAULT_BASE_URL

        client = AsyncOpenAI(
            api_key=os.environ.get("XAI_API_KEY"),
            base_url=os.environ.get("XAI_BASE_URL", DEFAULT_BASE_URL),
        )

        chat_system_prompt = (
            "你是一个工控安全领域的 AI 助手，专注于工业控制系统（ICS/SCADA）安全。\n\n"
            "你的职责：\n"
            "1. 回答用户关于工控安全的问题，包括安全告警分析、攻击链推断、防护策略等\n"
            "2. 对一般性问候和闲聊做出友好、自然的回应\n"
            "3. 当用户询问当前安全态势时，基于\"当前系统状态\"中的真实数据进行分析和回答\n"
            "4. 可以引用 MITRE ATT&CK for ICS 框架来解释攻击手法\n\n"
            "回复风格：\n"
            "- 使用自然的对话语言，支持 Markdown 格式\n"
            "- 对一般性对话友好回应，不要强行往安全告警上靠\n"
            "- 对专业问题给出详细、专业的回答\n"
            "- 引用具体数据时注明来源（如\"根据系统数据...\"）\n"
            "- 回复应简洁清晰，避免冗余"
        )

        if system_context:
            chat_system_prompt += f"\n\n--- 当前系统状态 ---\n{system_context}"

        messages = [{"role": "system", "content": chat_system_prompt}]
        messages += [{"role": m.role, "content": m.content} for m in body.messages]

        model = os.environ.get("XAI_MODEL", DEFAULT_MODEL)
        response = await client.chat.completions.create(
            model=model,
            messages=messages,
            temperature=0.5,
        )

        reply = response.choices[0].message.content or ""
        return {"reply": reply}
    except HTTPException:
        raise
    except Exception as exc:
        logger.exception("Chat failed")
        raise HTTPException(status_code=500, detail="Chat service temporarily unavailable") from exc


@router.get("/chains")
async def list_chains():
    """Query attack chain list."""
    conn = get_db()
    raw_chains = conn.execute(
        "SELECT c.* FROM attack_chains c ORDER BY c.created_at DESC LIMIT 50"
    ).fetchall()

    chains = []
    for c in raw_chains:
        c_dict = dict(c)
        try:
            alert_ids = json.loads(c_dict.get("alert_ids") or "[]")
        except (json.JSONDecodeError, TypeError):
            alert_ids = []

        alerts = []
        if alert_ids:
            ph = ",".join("?" for _ in alert_ids)
            alerts = [dict(a) for a in conn.execute(
                f"SELECT id, title, severity, src_ip FROM alerts WHERE id IN ({ph})", alert_ids
            ).fetchall()]

        decisions = [dict(d) for d in conn.execute(
            "SELECT * FROM decisions WHERE attack_chain_id = ? ORDER BY created_at DESC, id DESC",
            (c_dict["id"],),
        ).fetchall()]

        alert_count = len(alert_ids)
        pending_decisions = [d for d in decisions if d.get("status") == "pending"]
        status = "new" if not decisions else ("pending" if pending_decisions else "resolved")

        primary = decisions[0] if decisions else None
        risk_level = (primary or {}).get("risk_level") or c_dict.get("risk_level", "medium")
        recommendation = (primary or {}).get("recommendation")
        action_type = (primary or {}).get("action_type")
        decision_status = (primary or {}).get("status")

        for d in decisions:
            d["action"] = d.get("recommendation")

        chains.append({
            **c_dict,
            "alerts": alerts,
            "decisions": decisions,
            "alert_count": alert_count,
            "status": status,
            "risk_level": risk_level,
            "recommendation": recommendation,
            "action_type": action_type,
            "decision_status": decision_status,
        })

    return {"chains": chains}


@router.patch("/decisions/{decision_id}")
async def decision_feedback(decision_id: int, body: DecisionFeedbackRequest):
    """Human feedback: accept or reject a decision."""
    status = body.status

    conn = get_db()
    # Atomic update: only succeeds if decision is still pending
    cursor = conn.execute(
        "UPDATE decisions SET status = ? WHERE id = ? AND status = ?",
        (status, decision_id, "pending"),
    )
    if cursor.rowcount == 0:
        existing = conn.execute("SELECT id, status FROM decisions WHERE id = ?", (decision_id,)).fetchone()
        if not existing:
            raise HTTPException(status_code=404, detail="Decision not found")
        raise HTTPException(
            status_code=400,
            detail=f"Decision already processed (status: {existing['status']})",
        )

    # Update related alert statuses
    try:
        decision = conn.execute(
            "SELECT attack_chain_id FROM decisions WHERE id = ?", (decision_id,)
        ).fetchone()
        if decision:
            chain = conn.execute(
                "SELECT alert_ids FROM attack_chains WHERE id = ?", (decision["attack_chain_id"],)
            ).fetchone()
            if chain and chain["alert_ids"]:
                try:
                    alert_ids = json.loads(chain["alert_ids"])
                    new_alert_status = "resolved" if status == "accepted" else "analyzed"
                    for aid in alert_ids:
                        conn.execute("UPDATE alerts SET status = ? WHERE id = ?", (new_alert_status, aid))
                except (json.JSONDecodeError, TypeError):
                    pass

        # Write audit log
        conn.execute(
            "INSERT INTO audit_logs (trace_id, alert_id, event_type, data) VALUES (?, ?, ?, ?)",
            (f"decision-{decision_id}", "", "decision_feedback",
             json.dumps({"decision_id": decision_id, "status": status})),
        )
    except Exception:
        logger.exception("Failed to update related alerts")

    conn.commit()
    return {"updated": True}


# ---------------------------------------------------------------------------
# System context builder (ported from Express)
# ---------------------------------------------------------------------------

def _build_system_context(conn) -> str:
    try:
        parts: list[str] = []

        stats = conn.execute("""
            SELECT
                COUNT(*) as total,
                SUM(CASE WHEN status = 'open' THEN 1 ELSE 0 END) as open_count,
                SUM(CASE WHEN status = 'analyzing' THEN 1 ELSE 0 END) as analyzing_count,
                SUM(CASE WHEN severity = 'critical' THEN 1 ELSE 0 END) as critical_count,
                SUM(CASE WHEN severity = 'high' THEN 1 ELSE 0 END) as high_count,
                SUM(CASE WHEN severity = 'medium' THEN 1 ELSE 0 END) as medium_count,
                SUM(CASE WHEN severity = 'low' THEN 1 ELSE 0 END) as low_count
            FROM alerts
        """).fetchone()

        if stats and stats["total"] > 0:
            parts.append(f"告警概览: 共{stats['total']}条告警, 待处理{stats['open_count']}条, 分析中{stats['analyzing_count']}条")
            parts.append(f"严重度分布: 严重{stats['critical_count']} / 高{stats['high_count']} / 中{stats['medium_count']} / 低{stats['low_count']}")
        else:
            parts.append("告警概览: 当前系统无告警记录")

        recent = conn.execute(
            "SELECT id, title, severity, source, src_ip, dst_ip, status, created_at FROM alerts ORDER BY created_at DESC LIMIT 5"
        ).fetchall()
        if recent:
            parts.append("\n最近告警:")
            for a in recent:
                parts.append(
                    f"  - [{a['severity']}] {a['title']} (来源:{a['source']}, "
                    f"{a['src_ip'] or '?'}->{a['dst_ip'] or '?'}, 状态:{a['status']})"
                )

        chains = conn.execute(
            "SELECT id, name, stage, confidence FROM attack_chains ORDER BY created_at DESC LIMIT 3"
        ).fetchall()
        if chains:
            parts.append("\n最近攻击链:")
            for c in chains:
                parts.append(f"  - {c['name']} (阶段:{c['stage']}, 置信度:{c['confidence']})")

        asset_count = conn.execute("SELECT COUNT(*) as cnt FROM assets").fetchone()
        if asset_count and asset_count["cnt"] > 0:
            parts.append(f"\n资产: 共{asset_count['cnt']}台受监控资产")

        return "\n".join(parts)
    except Exception:
        logger.exception("Failed to build system context")
        return ""
